vm/qemu/qemu.go

```diff
@@ -364,9 +418,12 @@ func (inst *instance) Close() {
 
 func (inst *instance) boot() error {
 	inst.port = vmimpl.UnusedTCPPort()
+	inst.monport = vmimpl.UnusedTCPPort()
 	args := []string{
 		"-m", strconv.Itoa(inst.cfg.Mem),
 		"-smp", strconv.Itoa(inst.cfg.CPU),
+		"-chardev", fmt.Sprintf("socket,id=SOCKSYZ,server,nowait,host=localhost,port=%v", inst.monport),
```

Comment:
Interesting. What other types of info are available there?
There is #1217 for doing this. + @ajdlinux @mpe 
Is it possible to obtain kernel core dump using this? #491

Reply:
This allows access to the machine state as QEMU sees it and also direct access to the VM's RAM so you can dump the entire RAM or (as an example) I use it to dump dmesg (System.map tells where it is so I do not need to dump the entire RAM). The dumping is quite slow though, 1MB takes more than 5seconds.

The current output is something like this:
```
21:49:25 executing program 0:
openat$kvm(0xffffffffffffff9c, &(0x7f0000000040)='/dev/kvm\x00', 0x0, 0x0)
openat$kvm(0xffffffffffffff9c, &(0x7f0000000000)='/dev/kvm\x00', 0x125040, 0x0)
ioctl$KVM_CREATE_VM(0xffffffffffffffff, 0x2000ae01, 0x0)
openat$kvm(0xffffffffffffff9c, &(0x7f0000000080)='/dev/kvm\x00', 0x420600, 0x0)

BUG: Unable to handle kernel data access on write at 0x5cd800a0
Faulting instruction address: 0xc00000000029326c
cpu 0x4: Vector: 300 (Data Access) at [c000000005367800]
    pc: c00000000029326c: lockdep_init_map_waits+0x3c/0x2d0
    lr: c0000000002993a8: __raw_spin_lock_init+0x58/0xa0
    sp: c000000005367a90
   msr: 800000000280b033
   dar: 5cd800a0
 dsisr: 42000000
  current = 0xc0000000052a9900
  paca    = 0xc00000003fff9300	 irqmask: 0x03	 irq_happened: 0x01
    pid   = 958, comm = syz-executor.4
Linux version 5.7.0-le-guest_syzkaller_a+fstn1 (aik@fstn1-p1) (gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0, GNU ld (GNU Binutils for Ubuntu) 2.30) #466 SMP Mon Jun 22 11:07:21 AEST 2020
enter ? for help
[c000000005367b10] c0000000002993a8 __raw_spin_lock_init+0x58/0xa0
[c000000005367b60] c0000000002c5b20 init_srcu_struct_fields+0x430/0x630
[c000000005367c30] c0000000001598c4 kvm_dev_ioctl+0x234/0xe10
Connection to localhost closed by remote host.

VM DIAGNOSIS:
07:49:25  Registers:
NIP c00000000011cf6c   LR c000000001411e14 CTR c000000003570000 XER 0000000020040000 CPU#0
MSR 800000000280b033 HID0 0000000000000000  HF 8000000000000000 iidx 3 didx 3
TB 00000000 00000000 DECR 0
GPR00 0000000024000828 c000000002623c80 c000000002624200 0000000000000000
GPR04 0000000000000006 0000000000000001 00000016628920c8 000000005cd80000
GPR08 0000000000000000 c000000002597800 0000000000000000 c000000002657a88
GPR12 0000000000000000 c000000003570000 0000000000000000 0000000002f71a00
GPR16 0000000002f71d40 0000000002f71b70 0000000002f72108 0000000002fec398
GPR20 fffffffffffffffd 0000000000000888 0000000000000001 c00000000253a5f8
GPR24 0000000000000000 c00000000256b1b8 00000053e7ea2990 0000000000000001
GPR28 c00000000266fb18 0000000000000001 c000000001e274a0 c000000001e274a8
CR 44000828  [ G  G  -  -  -  L  E  L  ]             RES ffffffffffffffff
FPR00 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR04 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR08 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR12 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR16 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR20 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR24 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPR28 0000000000000000 0000000000000000 0000000000000000 0000000000000000
FPSCR 0000000000000000
 SRR0 c00000000011cf6c  SRR1 8000000002803033    PVR 00000000004e1202 VRSAVE 0000000000000000
SPRG0 0000000000000000 SPRG1 c000000003570000  SPRG2 00000000110d7760  SPRG3 0000000000000000
SPRG4 0000000000000000 SPRG5 0000000000000000  SPRG6 0000000000000000  SPRG7 0000000000000000
HSRR0 0000000000000000 HSRR1 0000000000000000
 CFAR 0000000000000000
 LPCR 0000000003d60413
 PTCR 0000000000000000   DAR 00007fff91d8f218  DSISR 0000000042000000
```

Reply:
@dvyukov Using the qemu monitor you can inject an NMI with "nmi". On powerpc that becomes a system reset exception which will then trigger kdump if it's configured. So yes you could trigger a kernel dump that way.

There are bugs where kdump won't work, ie. if the kernel is sufficiently broken. So dumping registers using qemu first would be good IMO.